x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/HashJoinExec.java

```diff
@@ -49,10 +49,11 @@ public HashJoinExec(
         this.output = output;
     }
 
+    @SuppressWarnings("unchecked")
     public HashJoinExec(PlanStreamInput in) throws IOException {
         super(Source.readFrom(in), in.readPhysicalPlanNode());
         this.joinData = new LocalSourceExec(in);
-        this.matchFields = in.readNamedWriteableCollectionAsList(NamedExpression.class);
+        this.matchFields = (List<Attribute>) (List) in.readNamedWriteableCollectionAsList(NamedExpression.class);
```

Comment:
Since we haven't cut a release of with this and we don't allow these constructs in serverless yet I think you can change this to `in.readNamedWriteableCollectionAsList(Attribute.class)`. There isn't any code that'll ever send anything that isn't an `Attribute`.

Also, if these were always `Attribute` subclasses in practice then this'd be safe here too - it's just on the read side.